[Bug Report] shutdown crash before start signal thread

### Overview of the Issue

When the start_binlog_archive operation fails, the process exits with a core dump.
2024-11-20T03:48:51.613899Z 1 [System] [MY-013576] [InnoDB] InnoDB initialization has started.
2024-11-20T03:48:57.262760Z 1 [System] [MY-013577] [InnoDB] InnoDB initialization has ended.
2024-11-20T03:48:59.935615Z 0 [Warning] [MY-000000] [Server] SE: This node is not the owner of the lease lock, waiting...
2024-11-20T03:49:00.607671Z 0 [Warning] [MY-000000] [Server] this node becomes the lease lock owner! new lease time:1732074552325
2024-11-20T03:49:00.952823Z 0 [Warning] [MY-000000] [Server] SE: This node becomes the owner of the lease lock, go on
[2024-11-20 03:49:00.968544] SYSTEM [628] [SE] dump([file_name.cc:43](http://file_name.cc:43/)) SMARTENGINE FILES SUMMARY:
[2024-11-20 03:49:00.972083] WARN   [628] [SE] dump_dir([file_name.cc:198](http://file_name.cc:198/)) fail to parse file name, ret=21, file_name="SMARTENGINE_PERSISTENT_CACHE", file_number=0, file_type=2
[2024-11-20 03:49:00.974475] SYSTEM [628] [SE] dump_dir([file_name.cc:237](http://file_name.cc:237/)) Files Summary in directory: smartengine, sst_file_count = 0, wal_file_count = 2, manifest_file_count = 3, checkpoint_file_count = 1, temp_file_count = 0
[2024-11-20 03:49:01.014033] SYSTEM [628] [SE] load_checkpoint([storage_logger.cc:813](http://storage_logger.cc:813/)) success to load checkpoint, checkpoint_name.c_str()="0.checkpoint", log_number=0
[2024-11-20 03:49:01.060751] SYSTEM [628] [SE] replay_after_ckpt([storage_logger.cc:896](http://storage_logger.cc:896/)) replay manifest file, log_file_number=1
[2024-11-20 03:49:01.072788] SYSTEM [628] [SE] replay_after_ckpt([storage_logger.cc:896](http://storage_logger.cc:896/)) replay manifest file, log_file_number=2
[2024-11-20 03:49:01.086268] SYSTEM [628] [SE] replay_after_ckpt([storage_logger.cc:896](http://storage_logger.cc:896/)) replay manifest file, log_file_number=3
[2024-11-20 03:49:01.601925] SYSTEM [628] [SE] replay_after_ckpt([storage_logger.cc:980](http://storage_logger.cc:980/)) success to replay after checkpoint, log_file_number=3
[2024-11-20 03:49:01.620175] SYSTEM [628] [SE] replay([storage_logger.cc:563](http://storage_logger.cc:563/)) success to replay manifest, log_file_number=4
[2024-11-20 03:49:01.623047] SYSTEM [628] [SE] before_replay_wal_files([db_impl_open.cc:416](http://db_impl_open.cc:416/)) max sequence number in recovery point, max_seq_in_rp=60
[2024-11-20 03:49:01.693040] SYSTEM [628] [SE] Open([db_impl_open.cc:1475](http://db_impl_open.cc:1475/)) success to open DBImpl,
[2024-11-20 03:49:01.693120] SYSTEM [750] [SE] bg_master_thread_func([db_impl.cc:745](http://db_impl.cc:745/)) smartengine Master Thread online,
2024-11-20T03:49:03.543942Z 0 [System] [MY-025082] [Repl] Plugin raft_replication reported: 'Force change consensus meta successfully'
2024-11-20T03:49:03.545724Z 0 [System] [MY-025028] [Repl] Plugin raft_replication reported: 'Consensus recovery start apply index, recover status: 0, first_index: 1, start apply index: 0, committed HWM index: 0, applier index: 0.'
2024-11-20T03:49:03.547043Z 0 [System] [MY-025029] [Repl] Plugin raft_replication reported: 'Consensus recovery start apply index, recovered index: 1.'
2024-11-20T03:49:03.550847Z 0 [System] [MY-025030] [Repl] Plugin raft_replication reported: 'Consensus recovery gtid sets finished.'
2024-11-20T03:49:03.564324Z 0 [Warning] [MY-010068] [Server] CA certificate ca.pem is self signed.
2024-11-20T03:49:03.566015Z 0 [System] [MY-013602] [Server] Channel mysql_main configured to support TLS. Encrypted connections are now supported for this channel.
2024-11-20T03:49:03.746546Z 0 [Warning] [MY-025123] [Server] Binlog archive startup: error 28, failed to create archive dir /tmp/binlog/.
2024-11-20T03:49:03.747796Z 0 [System] [MY-025123] [Server] Binlog archive startup: start binlog archive.
2024-11-20T03:49:03.746611Z 8 [System] [MY-025074] [Repl] Plugin raft_replication reported: 'Consensus applier thread start, recover status: 0, consensus start apply index: 0, rli consensus index: 0.'
2024-11-20T03:49:03.756601Z 8 [System] [MY-025077] [Repl] Plugin raft_replication reported: 'Consensus applier start position, log file: './binlog.000001', pos: 251, apply index: 0.'
2024-11-20T03:49:04.544476Z 0 [System] [MY-025124] [Server] Binlog archive create object store: provider=aws region=cn-northwest-1 endpoint= bucket=wesql-ljc-test use_https=false .
2024-11-20T03:49:04.546320Z 13 [System] [MY-025120] [Server] Binlog archive: Binlog archive thread running in Data mode.
2024-11-20T03:49:04.547543Z 13 [System] [MY-025120] [Server] Binlog archive: create binlog archive update index worker.
2024-11-20T03:49:04.548572Z 13 [System] [MY-025122] [Server] Binlog archive update index worker: starting.
2024-11-20T03:49:04.546467Z 0 [System] [MY-025134] [Server] Snapshot archive: start snapshot archive ...
2024-11-20T03:49:04.550011Z 14 [System] [MY-025122] [Server] Binlog archive update index worker: thread running.
2024-11-20T03:49:04.550069Z 13 [System] [MY-025122] [Server] Binlog archive update index worker: started.
2024-11-20T03:49:04.553703Z 13 [System] [MY-025120] [Server] Binlog archive: create binlog archive worker.
2024-11-20T03:49:04.555193Z 13 [System] [MY-025121] [Server] Binlog archive worker 0: starting.
2024-11-20T03:49:04.550957Z 0 [Warning] [MY-025134] [Server] Snapshot archive: error 28, failed to create local snapshot dir /tmp/snapshot/
2024-11-20T03:49:04.557492Z 0 [ERROR] [MY-010119] [Server] Aborting
2024-11-20T03:49:04.556730Z 13 [System] [MY-025121] [Server] Binlog archive worker 0: started.
2024-11-20T03:49:04.559931Z 13 [System] [MY-025121] [Server] Binlog archive worker 1: starting.
2024-11-20T03:49:04.565668Z 16 [System] [MY-025121] [Server] Binlog archive worker 1: thread running.
2024-11-20T03:49:04.565838Z 13 [System] [MY-025121] [Server] Binlog archive worker 1: started.
2024-11-20T03:49:04.568686Z 13 [System] [MY-025121] [Server] Binlog archive worker 2: starting.
2024-11-20T03:49:04.556686Z 15 [System] [MY-025121] [Server] Binlog archive worker 0: thread running.
2024-11-20T03:49:04.570707Z 17 [System] [MY-025121] [Server] Binlog archive worker 2: thread running.
2024-11-20T03:49:04.570720Z 13 [System] [MY-025121] [Server] Binlog archive worker 2: started.
2024-11-20T03:49:04.573797Z 13 [System] [MY-025121] [Server] Binlog archive worker 3: starting.
2024-11-20T03:49:04.575302Z 18 [System] [MY-025121] [Server] Binlog archive worker 3: thread running.
2024-11-20T03:49:04.575382Z 13 [System] [MY-025121] [Server] Binlog archive worker 3: started.
2024-11-20T03:49:04.974545Z 0 [System] [MY-025012] [Repl] Plugin raft_replication reported: 'Plugin 'raft_replication' has been stopped.'
[2024-11-20 03:49:11.227849] [Default] Server 1 : Enter startElectionCallback
[2024-11-20 03:49:11.227849] [Default] Server 1 : Paxos state change from FOLL to CAND !!
[2024-11-20 03:49:11.227849] [Default] Server 1 : Epoch task currentEpoch(0) during requestVote
[2024-11-20 03:49:11.227849] [Default] Server 1 : Start new requestVote: new term(2)
[2024-11-20 03:49:11.227849] [Default] Server 1 : Paxos state change from CAND to LEDR !!
2024-11-20T03:49:11Z UTC - mysqld got signal 11 ;
Most likely, you have hit a bug, but this error can also be caused by malfunctioning hardware.
BuildID[sha1]=c7dd6bcbd1901b118b3687fe501f3018149c78bf
Thread pointer: 0x0
Attempting backtrace. You can use the following information to find out
where mysqld died. If you see no messages after this, something went
terribly wrong...
stack_bottom = 0 thread_stack 0x100000
/usr/sbin/mysqld(my_print_stacktrace(unsigned char const*, unsigned long)+0x44) [0x20d4834]
/usr/sbin/mysqld(print_fatal_signal(int)+0x374) [0x11a2cd4]
/usr/sbin/mysqld(handle_fatal_signal+0x98) [0x11a2d88]
linux-vdso.so.1(__kernel_rt_sigreturn+0) [0xffff88b63790]
/usr/sbin/mysqld(ConsensusFifoCacheManager::add_log_to_cache(unsigned long, unsigned long, unsigned long, unsigned char*, bool, unsigned int, unsigned long, bool)+0xac) [0x24b8f7c]
/usr/sbin/mysqld(consensus_append_log_entry(MYSQL_BIN_LOG*, ConsensusLogEntry&, unsigned long*, Relay_log_info*, bool)+0x404) [0x24af2f4]
/usr/sbin/mysqld(ConsensusLogManager::write_log_entry(ConsensusLogEntry&, unsigned long*, bool)+0xf4) [0x24be784]
/usr/sbin/mysqld(BLConsensusLog::appendWithCheck(alisql::LogEntry const&)+0x80) [0x24e3680]
/usr/sbin/mysqld(alisql::LocalServer::writeLog(alisql::LogEntry&)+0xa0) [0x306f6a0]
/usr/sbin/mysqld(alisql::Paxos::replicateLog_(alisql::LogEntry&, bool)+0x2c8) [0x3065af8]
/usr/sbin/mysqld(alisql::Paxos::becameLeader_()+0x370) [0x3068d60]
/usr/sbin/mysqld(alisql::Paxos::requestVote(bool)+0x21c) [0x3069f0c]
/usr/sbin/mysqld(alisql::ThreadTimer::callbackRunWeak(std::weak_ptr<alisql::ThreadTimer::CallbackBase>)+0x60) [0x307efc0]
/usr/sbin/mysqld(alisql::Service::CallbackImpl<std::_Bind<void (*(std::weak_ptr<alisql::ThreadTimer::CallbackBase>))(std::weak_ptr<alisql::ThreadTimer::CallbackBase>)> >::run()+0x58) [0x307f648]
/usr/sbin/mysqld(alisql::Service::onAsyncEvent(alisql::Service::Callback*)+0x18) [0x307c628]
/usr/sbin/mysqld(alisql::Service::process(easy_request_t*, void*)+0x680) [0x307ded0]
/usr/sbin/mysqld() [0x3098a60]
/usr/sbin/mysqld(ev_invoke_pending+0x54) [0x30a08b4]
/usr/sbin/mysqld() [0x3098db0]
/usr/sbin/mysqld(ev_run+0x248) [0x30a3ab8]
/usr/sbin/mysqld(easy_baseth_on_start+0x3c) [0x309918c]
/usr/sbin/mysqld() [0x307c5a4]
/lib64/libpthread.so.0(+0x7950) [0xffff889d3950]
/lib64/libc.so.6(+0x23e1c) [0xffff87d84e1c]
The manual page at http://dev.mysql.com/doc/mysql/en/crashing.html contains
information that should help you find out what is causing the crash.
Writing a core file

### Reproduction Steps

Start WeSQL when the disk fills up

### Binary Version

```sh
8.0.35-0.1.0_beta1
```

### Operating System and Environment details

```sh
linux/arm
```

### Log Fragments

```sh

```

--- a/plugin/raft_replication/consensus_log_manager.cc
+++ b/plugin/raft_replication/consensus_log_manager.cc
@@ -469,12 +469,18 @@ int ConsensusLogManager::truncate_log(uint64 consensus_index) {
   uint64 start_index = 0;
   my_off_t offset;
   std::string file_name;
+  bool use_new_created_thd = false;
   DBUG_TRACE;
 
   LogPluginErr(SYSTEM_LEVEL, ER_CONSENSUS_LOG_MANAGER_LOG_OPS, "truncate",
                consensus_index);
 
   prefetch_manager->stop_prefetch_threads();
+
+  if (current_thd == nullptr) {
+    use_new_created_thd = true;
+    current_thd = create_internal_thd();
+  }
 
   mysql_rwlock_rdlock(consensus_state_process.get_consensuslog_status_lock());
   mysql_rwlock_wrlock(&LOCK_consensuslog_truncate);
@@ -528,6 +534,11 @@ int ConsensusLogManager::truncate_log(uint64 consensus_index) {
   mysql_rwlock_unlock(&LOCK_consensuslog_truncate);
   mysql_rwlock_unlock(consensus_state_process.get_consensuslog_status_lock());
 
+  if (use_new_created_thd) {
+    destroy_internal_thd(current_thd);
+    current_thd = nullptr;
+  }
+
   if (error) abort();
 
   prefetch_manager->start_prefetch_threads();
@@ -539,9 +550,15 @@ int ConsensusLogManager::purge_log(uint64 consensus_index) {
   std::string file_name;
   uint64 start_index;
   uint64 purge_index = 0;
+  bool use_new_created_thd = false;
   DBUG_TRACE;
 
   if (consensus_index == 0) return 0;
+
+  if (current_thd == nullptr) {
+    use_new_created_thd = true;
+    current_thd = create_internal_thd();
+  }
 
   mysql_rwlock_rdlock(consensus_state_process.get_consensuslog_status_lock());
 
@@ -617,6 +634,11 @@ int ConsensusLogManager::purge_log(uint64 consensus_index) {
   }
 
   mysql_rwlock_unlock(consensus_state_process.get_consensuslog_status_lock());
+
+  if (use_new_created_thd) {
+    destroy_internal_thd(current_thd);
+    current_thd = nullptr;
+  }
   return error;
 }
 

--- a/plugin/raft_replication/observer_server_state.cc
+++ b/plugin/raft_replication/observer_server_state.cc
@@ -61,23 +61,10 @@ int consensus_replication_after_recovery(Server_state_param *) {
   /* If the plugin is not running, return failed. */
   if (!plugin_is_consensus_replication_running()) return 1;
 
+  if (start_consensus_replication()) return 1;
+
   if (!opt_initialize) {
-    /* Init replica construct for consensus channel */
-    if (init_consensus_replica()) return -1;
-
-    if (consensus_state_process.start_consensus_state_change_thread())
-      return -1;
-
-    /* Start consensus service */
-    if (consensus_state_process.init_service()) return -1;
-
-    if (consensus_log_manager.start_consensus_commit_advance_thread())
-      return -1;
-
-    /* Start consensus apply threads */
-    if (!opt_cluster_log_type_instance && start_consensus_replica()) return -1;
-
-    // purge logs after start applier thread
+    // purge logs after start consensus replication
     MYSQL_BIN_LOG *log =
         &consensus_state_process.get_relay_log_info()->relay_log;
     if (DBUG_EVALUATE_IF("expire_logs_always_at_start", false, true))
@@ -96,6 +83,9 @@ int consensus_replication_before_handle_connection(Server_state_param *) {
 
   /* If the plugin is not running, return failed. */
   if (!plugin_is_consensus_replication_running()) return 1;
+
+  /* Start consensus apply threads */
+  if (!opt_cluster_log_type_instance && start_consensus_replica()) return -1;
 
   if (!opt_initialize) rpl_consensus_set_ready();
 
@@ -116,22 +106,12 @@ int consensus_replication_after_server_shutdown(Server_state_param *) {
   /* If the plugin is not running, return failed. */
   if (!plugin_is_consensus_replication_running()) return 1;
 
-  if (!opt_initialize) {
-    /* Stop consensus commit advance */
-    consensus_log_manager.stop_consensus_commit_advance_thread();
-    /* Stop consensus state change */
-    consensus_state_process.stop_consensus_state_change_thread();
-    /* Stop consensus relica */
-    if (!opt_cluster_log_type_instance) end_consensus_replica();
-    /* Stop consensus service */
-    rpl_consensus_shutdown();
-    rpl_consensus_cleanup();
-  }
+  stop_consensus_replication();
 
-  /* Close mysql_bin_log after shutdown */
+  /* Close mysql_bin_log after stop consensus replication service */
   consensus_state_process.get_binlog()->close(LOG_CLOSE_INDEX,
-                                            true /*need_lock_log=true*/,
-                                            true /*need_lock_index=true*/);
+                                              true /*need_lock_log=true*/,
+                                              true /*need_lock_index=true*/);
 
   return 0;
 }

--- a/plugin/raft_replication/plugin.cc
+++ b/plugin/raft_replication/plugin.cc
@@ -194,24 +194,52 @@ err:
   return error;
 }
 
-int plugin_consensus_replication_stop(char **) {
+int start_consensus_replication() {
+  if (opt_initialize) return 0;
+
+  /* Init replica construct for consensus channel */
+  if (init_consensus_replica()) return -1;
+
+  if (consensus_state_process.start_consensus_state_change_thread()) return -1;
+
+  /* Start consensus service */
+  if (consensus_state_process.init_service()) return -1;
+
+  if (consensus_log_manager.start_consensus_commit_advance_thread()) return -1;
+
+  return 0;
+}
+void stop_consensus_replication() {
+  if (opt_initialize) return;
+
+  /* Stop consensus commit advance */
+  consensus_log_manager.stop_consensus_commit_advance_thread();
+  /* Stop consensus state change */
+  consensus_state_process.stop_consensus_state_change_thread();
+  /* Stop consensus relica */
+  if (!opt_cluster_log_type_instance) end_consensus_replica();
+  /* Stop consensus service */
+  rpl_consensus_shutdown();
+  rpl_consensus_cleanup();
+}
+
+int plugin_consensus_replication_stop() {
   DBUG_TRACE;
 
   Checkable_rwlock::Guard g(*lv.plugin_running_lock,
                             Checkable_rwlock::WRITE_LOCK);
 
-  if (!plugin_is_consensus_replication_running()) {
-    return 0;
-  }
+  if (!plugin_is_consensus_replication_running()) return 0;
+
+  if (!opt_initialize && !rpl_consensus_is_shutdown())
+    stop_consensus_replication();
 
   lv.plugin_stop_lock->wrlock();
   LogPluginErr(INFORMATION_LEVEL, ER_CONSENSUS_RPL_IS_STOPPING);
 
   lv.consensus_replication_running = false;
 
-  if (!opt_initialize) {
-    finalize_registry_handler();
-  }
+  if (!opt_initialize) finalize_registry_handler();
 
   lv.plugin_stop_lock->unlock();
   LogPluginErr(SYSTEM_LEVEL, ER_CONSENSUS_RPL_IS_STOPPED);

--- a/plugin/raft_replication/plugin.h
+++ b/plugin/raft_replication/plugin.h
@@ -37,7 +37,10 @@ bool plugin_is_consensus_replication_running();
 // Plugin public methods
 int plugin_consensus_replication_init(MYSQL_PLUGIN plugin_info);
 int plugin_consensus_replication_deinit(void *p);
-int plugin_consensus_replication_stop(char **error_message = nullptr);
+int plugin_consensus_replication_stop();
+
+int start_consensus_replication();
+void stop_consensus_replication();
 
 bool acquire_transaction_control_services();
 bool release_transaction_control_services();

--- a/plugin/raft_replication/rpl_consensus.cc
+++ b/plugin/raft_replication/rpl_consensus.cc
@@ -119,16 +119,20 @@ void rpl_consensus_set_ready() {
 bool rpl_consensus_is_ready() { return rpl_consensus_ready; }
 
 void rpl_consensus_shutdown() {
-  if (rpl_consensus_inited && consensus_ptr && !consensus_ptr->isShutdown())
+  if (rpl_consensus_inited && !consensus_ptr->isShutdown()) {
     consensus_ptr->shutdown();
+  }
 }
 
 void rpl_consensus_cleanup() {
   if (consensus_ptr) delete consensus_ptr;
+  rpl_consensus_inited = false;
   consensus_ptr = nullptr;
 }
 
-bool rpl_consensus_is_shutdown() { return consensus_ptr->isShutdown(); }
+bool rpl_consensus_is_shutdown() {
+  return (!rpl_consensus_inited || consensus_ptr->isShutdown());
+}
 
 bool rpl_consensus_is_leader() {
   return (consensus_ptr->getState() == alisql::Paxos::LEADER);

